src/twisted/internet/asyncfs.py

```diff
@@ -0,0 +1,709 @@
+        @rtype: L{int}
+        """
+
+    def getAttrs() -> "Deferred":
+        """
+        Return the attributes for the file.
+
+        This method returns a dictionary in the same format as the attrs
+        argument to L{openFile} or a L{Deferred} that is called back with same.
+        """
+
+    def setAttrs(attrs: dict) -> "Deferred":
+        """
+        Set the attributes for the file.
+
+        This method returns when the attributes are set or a Deferred that is
+        called back when they are.
+
+        @param attrs: a dictionary in the same format as the attrs argument to
+        L{openFile}.
+        """
+
+    def fileno() -> int:
+        """
+        @return: the underlying file descriptor
+        @rtype: L{int)
+        """
+
+    def send(
+        consumer: "IConsumer", start: int = 0, chunkSize: int = DEFAULT_CHUNK_SIZE
+    ) -> "Deferred":
+        """
+        Produce the contents of the file to the given consumer.
+
+        @type consumer: C{IConsumer}
+        @param start: offset to begin reading from
+        @param chunkSize: size, in bytes, of each read
+        @return: A Deferred which fires when the file has been
+        consumed completely.
+
+        adapted from L{twisted.protocols.ftp.IReadFile}
+        """
+
+    producer = Attribute("a read-only IPushProducer object available during send()")
+
+    def receive(append: bool = False) -> "IConsumer":
+        """
+        @param append: True if append to end of file
+        @return: A C{IConsumer} which is used to write data
+
+        adapted from L{twisted.protocols.ftp.IWriteFile}
+        """
+
+
+class AccessViolation(Exception):
+    """thrown when a caller tries to break access restrictions"""
+
+
+@implementer(IFile)
+class _ThreadFile:
+    """
+    a file accessed asynchronously using threads. Should only be created
+    by L{ThreadVfs.openFile}
+    """
+
+    def __init__(
+        self, fso: "ThreadFs", filename: str, flags: int, attrs: Optional[dict]
+    ) -> None:
+        """
+        @param fso: the parent filesystem object
+        @type fso: L{IFilesystem}
+        @param filename: the path to open
+        @type filename: L{str}
+        @param flags: opening flags, passed to L{os.open}
+        @type flags: L{int}
+        @param attrs: a dictionary in the same format as the attrs argument to
+        L{openFile}.
+        @type attrs: L{dict}
+        """
+        self.fso = fso
+        self._write_consumer: Optional["_ThreadFileConsumer"] = None
+        self._producer = _ThreadFileProducer()
+        if attrs is None:
+            attrs = {}
+        if "permissions" in attrs and not self.fso.read_only:
+            mode = attrs["permissions"]
+            del attrs["permissions"]
+        else:
+            mode = 0o644
+        if self.fso.read_only:
+            flags = os.O_RDONLY
+        elif flags == 0:
+            flags = os.O_RDWR
+        fd = os.open(filename, flags, mode)
```

Comment:
At first I thought this I/O was happening in the wrong thread.  Eventually I noticed that `_ThreadFile` is called in a non-reactor thread.  So that's fine, I guess.  It would be good to document that `_ThreadFile.__init__` runs in a non-reactor thread.

Also, I/O inside `__init__` is kind of an anti-pattern.  I suggest a classmethod constructor that does the I/O and leaves `__init__` to just setting attributes.

Speaking of just setting attributes, how about some `@attr.s` usage here?  It would probably shave down the linecount.